### ISSUE TITLE
Pre-release versions without channels

### ISSUE THREAD

**Issue by lukiffer** [semantic-release/semantic-release#1101]
## New feature motivation

We've historically versioned our packages using pre-releases (e.g. `1.0.0-beta.1`) and have always used the prerelease type as the release channel. It appears this is also the default behavior in the beta release (see https://github.com/semantic-release/semantic-release/blob/beta/lib/branches/normalize.js#L23)

This however also adds the channel to git tags (e.g. `1.0.0-beta.1@beta`). This is not only redundant in the case where the prerelease type is always the channel, but also incongruent with our historical versions. As a result, existing pre-release versions/tags aren't found because they're parsed as `channel===undefined` when searching for previous tags:

`context` from the plugin pipeline:

```js
  branches:
   [ { channel: undefined,
       tags: [Array],
       type: 'release',
       name: 'master',
       range: '>=1.0.0',
       accept: [Array] },
     { channel: undefined,
       type: 'prerelease',
       name: 'beta',
       prerelease: 'beta',
       tags: [Array] } ],
  branch:
   { channel: 'beta',
     type: 'prerelease',
     name: 'beta',
     prerelease: 'beta',
     tags: [ [Object], [Object] ] }
```

So a pre-existing tag `1.0.0-beta.1` isn't considered because it's actually looking for `1.0.0-beta.1@beta` because the normalization process is, by default, setting the channel to the branch name if it `isNil`. 

## New feature description

It would be convenient if the `isNil` check was removed; thus, if channels were intended to be used, that they would need to be explicitly specified in the config.

## New feature implementation

In _very_ limited local testing, the `isNil` check could be removed without ill effect, other than potentially breaking build/release chains that rely on the implicit use of the branch name as the channel.

**Comment by lukiffer** [comment]
After looking into this a bit further, there's a bit more that would have to change other than simply removing the `isNil` ternary:

It appears that the channel criteria is (as expected) being used to determine the latest release, but it is not using the prerelease type (e.g. an `alpha` branch with tag `1.1.0-alpha.3` being merged into a `beta` branch (with `{"prerelease": "beta"}`, and no `beta` tag produces:

```
[11:44:54] [semantic-release] › ℹ  Found git tag 1.1.0-alpha.3 associated with version 1.1.0-alpha.3 on branch stage
[11:44:54] [semantic-release] › ℹ  Found 0 commits since last release
[11:44:54] [semantic-release] › ℹ  Start step "analyzeCommits" of plugin "@semantic-release/commit-analyzer"
[11:44:54] [semantic-release] [@semantic-release/commit-analyzer] › ℹ  Analysis of 0 commits complete: no release
[11:44:54] [semantic-release] › ✔  Completed step "analyzeCommits" of plugin "@semantic-release/commit-analyzer"
[11:44:54] [semantic-release] › ℹ  There are no relevant changes, so no new version is released.
```
One would expect that it would look for previous `x.x.x-beta.x` tags and in the case there were none would look back since the latest release version, thus producing `1.1.0-beta.1`.

**Comment by dominykas** [comment]
Was just about to create this same issue.

My motivation for not having the `@channel` is that by looking at npm version of a dep, you can't simply `git checkout vX.Y.Z-rc.N` - you have to know the original channel it was published in.

I see several other options to resolve this:

1. When tagging `vX.Y.Z-rc.N@channel`, _also_ add a tag at the same place for `vX.Y.Z-rc.N` (in fact, this is what I think I'll do as a stopgap). Possibly even make the channel-less version _the_ release (i.e. keep the tag@channel for reference)
2. Allow an explicit `channel: ""`. This would mean that when filtering tags in `get-last-release.js`, the `branch.channel === tag.channel` needs to account for `branch.channel === '' && isNil(tag.channel)`. This is a little ugly - I do think that allowing channel-less pre-releases by default.
3. Allow a custom extension point for plugins to categorize tags into channels or to override the last/next release version/tag.

**Comment by pvdlg** [comment]
Tags are the only state semantic-release can maintain between releases.
An important feature for us is to be able to move releases from channel to the other when the corresponding code is merged from one branch to the other. For example if `2.0.0` is released on `@next` when the corresponding code is merged to `master` we want to automatically add `2.0.0` to `@latest`.

In order to do that semantic-release need to know on which channel a given release has been made available. This is why we have to add the `@channel` to the tag name. So if `v2.0.0@next` exist but `v2.0.0` doesn't, then it means `2.0.0` is available on `@next`, it has been merged to `master` but has not yet been made available on `@latest`.

Without adding the `@channel` this feature could be only be possible with an external database maintaining a state of all the release for all repos. That would be obviously way more constraining.

Regarding the proposed options.

> 1. When tagging `vX.Y.Z-rc.N@channel`, _also_ add a tag at the same place for `vX.Y.Z-rc.N` (in fact, this is what I think I'll do as a stopgap). Possibly even make the channel-less version _the_ release (i.e. keep the tag@channel for reference)

That is not possible because we wouldn't be able to determine if the code associated with `vX.Y.Z-rc.N` was merged to `master` or not

> 2. Allow an explicit `channel: ""` / `channel: false`. This would mean that when filtering tags in `get-last-release.js`, the `branch.channel === tag.channel` needs to account for `branch.channel === '' && isNil(tag.channel)`. This is a little ugly - I do think that allowing channel-less pre-releases by default is probably better. Alternatively, the `""` or `false` could be normalized into `undefined`.

That is not possible because the absence of a channel is used for the default one (`master` branch / `@latest` dist-tag)

> 3. Allow a custom extension point for plugins to categorize tags into channels or to override the last/next release version/tag.

Externalizing that to a plugin would not change the fact that you need to maintain a state between releases and the only reasonable way to do that is tag naming.

**Comment by dominykas** [comment]
> That is not possible because the absence of a channel is used for the default one (`master` branch / `@latest` dist-tag)

I think this is the bit I don't understand. Why force the pre-releases into channels?

Channels are a tricky concept to explain and a lot of people don't need them or don't know how to use them, so for private use I'm keen to find any and all simplifications of the process.

There is the downside, that `npm i package-name`, if the package is not yet a dependency in `package.json`, will install the pre-release version (but it will update your `package.json` so you do have a chance to spot that you dragged in a pre-release). However, if the dep _is_ in the `package.json`, despite the fact that `@latest` is pointing to a pre-release, it will install a matching non-pre-release version.

I don't see the above as a problem for my use case (i.e. private packages), but I can see how that can be an issue in public open source.

Is that a big enough deal to completely prevent even the opt-in behavior? I think this is a reasonable simplification. Can it be hidden under an `allowUnsafeChannel` or smth option?

> Externalizing that to a plugin would not change the fact that you need to maintain a state between releases and the only reasonable way to do that is tag naming

Externalizing version retrieval and selection into a plugin takes the burden away from `semantic-release` altogether?

One more storage option is annotated tags, but I haven't considered that deeply.

**Comment by dominykas** [comment]
> That is not possible because we wouldn't be able to determine if the code associated with `vX.Y.Z-rc.N` was merged to `master` or not

Not sure I understand this? It's possible to check whether a commit, associated with a tag, is (or isn't) on a specific branch? Why is there a need to check whether it was merged or not? If it was merged - it's no longer a pre-release, therefore a regular `vX.Y.Z` will happen - `rc` has nothing to do with it?

**Comment by travi** [comment]
> Why force the pre-releases into channels?

I think its important to remember that a channel is always involved, but not specifying one means that the default one (`latest`) is used. `latest` does come with several conventions attached to it. You mentioned the fact that `npm install <package-name>` would install a pre-release rather than a stable version, but there are others as well. 

One that would be a big impact for users of tools like [greenkeeper](https://greenkeeper.io/) is that it sends PRs for new versions published to `latest`, but not to other channels. Personally, I automatically merge any greenkeeper PRs that pass all verification, so going automatically from a stable version to a new pre-release would not be desirable.

One of the many things that I find valuable about semantic-release is that it makes it harder to accidentally abuse conventions that would result in upset consumers while also making it easier to do the right thing, even for some of the complex areas like channels. The newest semantic-release versions have made pre-releases to channels so simple that I can't imagine it being valuable enough to simplify for an edge case that then makes it possible for users of semantic-release to opt-out of the protections that are currently in place.

**Comment by pvdlg** [comment]
The scenario we want to support is:
1. Push a commit to `master` => release version `1.0.0` on `@latest`
2. Create branch `next` from `master`
3. Push a commit to `next` => release version `1.1.0` on `@next`
4. Merge `next` into `master` => add release `1.1.0` to the `@latest` channel

How would you do the step 4 if you don't have the channel in the tag name?
Please describe the algorithm and the git command to run. As far as I can tell it's not possible.

**Comment by dominykas** [comment]
OK, so the use case is for the release on a channel without a `prerelease` suffix, right?

Is the channel still mandatory if the suffix is used?

Before I go into details, have these been discarded for some reasons I'm not aware of:
- having a `v1.1.0` and `v1.1.0@next` tags while on `next` branch, then when merged to master - add `v1.1.0@latest`, i.e. follow the same pattern, but the `v1.1.0` naked tag is still always there
- verify whether something is `latest` or not by actually looking at what `npm view` (or equivalent used) gives? i.e. if `npm view` says `latest` is `1.0.0`, and you have a `1.1.0` on `master` - update `latest` the dist tag?

**Comment by pvdlg** [comment]
Pre-release or not doesn't change the problem. We want to use that workflow between any branches.
We can't use `npm view` to figure out what's released were because we support any projects, including non Javascript ones and ones that doesn't release on npm.
This is why we support npm via plugin, as it's optional.

**Comment by dominykas** [comment]
Before I write things up, on more question.

> We can't use npm view 

Do channels matter at all if you're only releasing to git? i.e. does the presence of the tag without the channel (to indicate latest) make a difference?

I'll have to look into other publish targets, but I assume they also have a mechanism to determine current version in a channel, i.e. this is target specific?

**Comment by pvdlg** [comment]
The goal is to support any release target via plugins. Some might support channel, some other don't. Some might provide a way to find the channels on which a version is available. We can't assume anything regarding the target as we want to be generic enough to support anything.

**Comment by dominykas** [comment]
I'm sorry, I'm not including the Gitlab/non-npm flows, but I hope conceptually it's all the same, plus the source of truth is still the git repo, and that was the goal here, right?

> 1. Push a commit to `master` => release version `1.0.0` on `@latest`

Same as already implemented:

- `semantic-release` adds a `v1.0.0` tag
- it creates the GH release
- it publishes on npm with `latest` dist-tag

> 2. Create branch `next` from `master`

Same as already implemented:

- `semantic-release` sees no commits since `v1.0.0` and does nothing

> 3. Push a commit to `next` => release version `1.1.0` on `@next`

Almost the same as already implemented (except for the release name and two tags at the same commit):

- `semantic-release` sees new commits with `feat` since last release (`v1.0.0`) and therefore decides that the next release is a `v1.1.0`
- it knows that things on `next` branch are `prerelease: true`
- it creates a tag `v1.1.0` to indicate the version
- it creates a tag `v1.1.0@next` to indicate that the version is only on the `next` channel
- it creates a GH release with `prerelease: true` flag for `v1.1.0`
- it publishes on npm `next` channel

> 4. Merge `next` into `master` => add release `1.1.0` to the `@latest` channel

- `semantic-release` sees only a merge commit since `v1.1.0` - no new release is necessary
- it sees that at `v1.1.0`, there's also a `v1.1.0@next`, but the `master` is marked as `latest` channel, therefore it knows it needs to promote from `next` to `latest`
    - note: at `v1.0.0`, there was no corresponding `@next` tag, so no need to do anything out of the ordinary
    - it knows it because it can run `git tag --contains v1.1.0` to list all the tags at the same place (if it doesn't already know that by comparing that sha of every tag)
- it adds a `v1.1.0@latest` tag, so that it can check that in the future (there are now three tags - `v1.1.0`, `v1.1.0@next` and `v1.1.0@latest`, but the release name is still `v1.1.0`)
- it removes the prerelease flag from the GH release
- it updates npm `latest` dist tag to point to `v1.1.0`

---

Let's play this through with what the OP suggested, i.e. no channels

- discussion point: how to set this in configuration - an explicit `channel: false` for the `beta` branch or something else

> 1. Push a commit to `master` => release version `1.0.0` on `@latest`

Same as above

> 2. Create branch `beta` from `master`

Same as above

> 3. Push a commit to `beta`

- `semantic-release` sees new commits with `feat` since last release (`v1.0.0`) and therefore decides that the next release is a `v1.1.0`
- it knows that things on `beta` branch are `prerelease: 'beta'`, without a channel
- it creates a tag `v1.1.0-beta.1`
- it creates a GH pre-release for `v1.1.0-beta.1`
- it publishes on npm
- it immediately restores the `latest` dist-tag to what it was before this publish, i.e. back to `v1.0.0`
    - ideally publishing a version with an explicitly empty dist-tag would be a feature in npm

> 4. Merge `beta` into `master`

- `semantic-release` sees a merge commit and the `feat` commit since `v1.0.0`
- it ignores a `v1.1.0-beta.1` tag because we're on the `master` branch
    - discussion point: should this have additionally had a `v1.1.0-beta.1@beta` if we had the `channel: 'beta'`?
- it tags a `v1.1.0`
- it creates a `v1.1.0` release on GH
- it publishes `v1.1.0` on npm with `latest` dist-tag
    - discussion point: if we had `channel: 'beta'`, should `beta` channel now get upgraded to `v1.1.0`? I think Chrome behaves roughly this way with their release channels?
    - discussion point: should the `v1.1.0` get added onto the same commit as `v1.1.0-beta.1` was, rather than on the merge commit?

---

tl;dr:

- if a commit has only a `vX.Y.Z`, then `vX.Y.Z` is on `['latest']`
- if a commit has a `vX.Y.Z` and `vX.Y.Z@channel1`, then `vX.Y.Z` is on `['channel1']`
- if a commit has a `vX.Y.Z`, a `vX.Y.Z@channel1`, a `vX.Y.Z@channel2`, a `vX.Y.Z@latest`, then `vX.Y.Z` is on `['latest', 'channel1', 'channel2']`
- if a commit has a `vX.Y.Z-pre.N`, then it is not on any channel, and that is fine
    - if this commit is newly merged on `master`, then it needs to be promoted to `vX.Y.Z` (either `HEAD` or the commit itself - debatable)
- if a commit has a `vX.Y.Z-pre.N@pre`, then it is on `['pre']`

**Comment by dominykas** [comment]
I'm happy to try to work on a PR if the above is acceptable? Not sure what the plans are for the v16 beta...

**Comment by felixfbecker** [comment]
The git equivalent of a version is a tag, and the git equivalent of a channel is a branch. sementic-release even determines the channel to release on through the branch. Why can't it determine which channel a version was released on through branches too? I.e. you can use git to determine if a certain tag (version) is reachable from a certain branch (channel). The relevant commands for this are `git branch --contains $TAG` (returns all branches/channels that a tag/version is reachable from/released on) and `git merge-base --is-ancestor $TAG $BRANCH` (checks if a tag is reachable from one specific branch).

Any reason why this wouldn't work?

**Comment by pvdlg** [comment]
> Any reason why this wouldn't work?

The reason that it wouldn't work is that when you merge a release branch to another the tag will be present in both branch but will not have been released on the branch targeted by the merge.
Basically there is now way to determine if a given tag has been merged in released or only merged.

**Comment by dominykas** [comment]
> will not have been released on the branch targeted by the merge

Not sure I follow? Can you give an example?

**Comment by pvdlg** [comment]
Fixed by #1368
